src/TorchSharp/Dataset.cs

```diff
@@ -13,21 +14,29 @@ public static partial class data
                 /// <summary>
                 /// Map-style data set
                 /// </summary>
-                public abstract class Dataset : Dataset<Dictionary<string, torch.Tensor>>
+                public abstract class Dataset : Dataset<Dictionary<string, Tensor>>,
+                    IDataset<IReadOnlyDictionary<string, Tensor>>
```

Comment:
As it commented here, due to covariation, `Dataset` should naturally be `IDataset<IReadOnlyDictionary<string, Tensor>>` (because it is `IDataset<Dictionary<string, Tensor>>`). However `FSharp.Examples` cannot be complied without this. I don't know why this happens.

I would suggest to remove this line, because it influences the detection of torch.utils.data.ConcatDataset (It doesn't know which type to use. `IReadOnlyDictionary` or `Dictionary` are both ok), but I failed to make the fsharp program work.

Reply:
A more radical idea is to remove `Dataset` and `IterableDataset` at all. Actually we don't need them. `DataLoader`s could work on any `IDataset<IReadOnlyDictionary<string, torch.Tensor>>` and `IDataset<IEnumerable<Tensor>>`.

By the way, `IterableDataset` occupied the position of PyTorch `IterableDataset`. #1353 That's also part of the reason why I wants to remove them at all.

Reply:
F# samples have to work.

Reply:
Hmm... I suppose so... However it doesn't https://dev.azure.com/dotnet/TorchSharp/_build/results?buildId=107828&view=results

Reply:
So. just to check -- with the current commits, the F# examples build, correct? I don't see any errors in the latest build.

Reply:
Yes I could be built by adding `: IDataset<IReadOnlyDictionary<string, Tensor>>`.